EncodingStreamingHighAvailability/HighAvailability/Services/JobOutputStatusSyncService.cs

```diff
@@ -0,0 +1,180 @@
+            Parallel.ForEach(instances, new ParallelOptions { MaxDegreeOfParallelism = 5 }, (mediaServiceInstanceHealthModel) =>
+            {
+                var allJobs = this.jobOutputStatusStorageService.ListByMediaServiceAccountNameAsync(mediaServiceInstanceHealthModel.MediaServiceAccountName, this.timeWindowToLoadJobsInMinutes).GetAwaiter().GetResult();
+                logger.LogInformation($"JobOutputStatusSyncService::SyncJobOutputStatusAsync loaded jobs history: instanceName={mediaServiceInstanceHealthModel.MediaServiceAccountName} count={allJobs.Count()}");
+                var aggregatedData = allJobs.GroupBy(i => i.JobName);
+                var jobsToRefresh = new List<JobOutputStatusModel>();
+                var uniqueJobCount = 0;
+
+                foreach (var jobData in aggregatedData)
+                {
+                    var finalStateReached = false;
+
+                    if (jobData.Any(j => j.JobOutputState == JobState.Finished) ||
+                        jobData.Any(j => j.JobOutputState == JobState.Error) ||
+                        jobData.Any(j => j.JobOutputState == JobState.Canceled))
+                    {
+                        finalStateReached = true;
+                    }
+
+                    if (!finalStateReached)
+                    {
+                        var lastUpdate = jobData.Max(j => j.EventTime);
+
+                        // if duration below max threshold, 
+                        if (currentTime.AddMinutes(-this.timeSinceLastUpdateToForceJobResyncInMinutes) > lastUpdate)
+                        {
+                            var jobToAdd = jobData.FirstOrDefault(j => !string.IsNullOrEmpty(j.TransformName));
+                            if (jobToAdd != null)
+                            {
+                                jobsToRefresh.Add(jobToAdd);
+                            }
+                        }
+                    }
+                    uniqueJobCount++;
+                }
+
+                this.RefreshJobOutputStatusAsync(mediaServiceInstanceHealthModel.MediaServiceAccountName, jobsToRefresh, uniqueJobCount, logger).GetAwaiter().GetResult();
+            });
+        }
+
+        private async Task RefreshJobOutputStatusAsync(string mediaServiceAccountName, IList<JobOutputStatusModel> jobOutputStatusModels, int totalNumberOfJobs, ILogger logger)
+        {
+            if (jobOutputStatusModels.Any())
+            {
+                var aggregatedDataByTransform = jobOutputStatusModels.GroupBy(i => i.TransformName);
+                foreach (var jobOutputStatusModelsPerTransform in aggregatedDataByTransform)
+                {
+                    var transformName = jobOutputStatusModelsPerTransform.Key;
+                    var count = jobOutputStatusModelsPerTransform.Count();
+
+                    if (count == 0)
+                    {
+                        continue;
+                    }
+
+                    if (count > totalNumberOfJobs / this.pageSize)
```

Comment:
Question 3: Need to validate if this is a good way to refresh missing status updates.

Reply:
Beverly provided info that job List operation is 8 times more expensive than single job Get operation to load job. Updated calculation to represent difference in resource cost.

Reply:
Beverly suggested to use 1:8 ratio between get and list calls. Implemented this change